File: Java/sorting/QuickSort.java
/**
 * Quick Sort Algorithm with Multiple Pivot Selection Strategies
 * 
 * Time Complexity: O(n log n) average case, O(n^2) worst case
 * Space Complexity: O(log n) due to recursion stack
 * 
 * Quick sort is a divide-and-conquer algorithm that picks an element as a pivot
 * and partitions the array around the pivot. The key process is the partition
 * operation.
 */
public class QuickSort {

  // Pivot selection strategies
  public enum PivotStrategy {
    LAST, // Last element as pivot
    FIRST, // First element as pivot
    MIDDLE, // Middle element as pivot
    RANDOM, // Random element as pivot
    MEDIAN_OF_THREE // Median of first, middle, and last elements
  }

  /**
   * Sorts an array using quick sort algorithm (ascending order).
   * Uses LAST element as pivot by default.
   * 
   * @param arr Array of integers to be sorted
   * @return Sorted array
   */
  public static int[] quickSort(int[] arr) {
    return quickSort(arr, PivotStrategy.LAST);
  }

  /**
   * Sorts an array using quick sort with specified pivot strategy.
   * 
   * @param arr      Array of integers to be sorted
   * @param strategy Pivot selection strategy
   * @return Sorted array
   */
  public static int[] quickSort(int[] arr, PivotStrategy strategy) {
    if (arr.length <= 1) {
      return arr;
    }

    quickSortHelper(arr, 0, arr.length - 1, strategy);
    return arr;
  }

  /**
   * Helper method for quick sort that recursively sorts the array.
   * 
   * @param arr      Array to be sorted
   * @param low      Starting index
   * @param high     Ending index
   * @param strategy Pivot selection strategy
   */
  private static void quickSortHelper(int[] arr, int low, int high, PivotStrategy strategy) {
    if (low < high) {
      // Partition the array and get the pivot index
      int pivotIndex = partition(arr, low, high, strategy);

      // Recursively sort elements before and after partition
      quickSortHelper(arr, low, pivotIndex - 1, strategy);
      quickSortHelper(arr, pivotIndex + 1, high, strategy);
    }
  }

  /**
   * Partitions the array around a pivot element.
   * 
   * @param arr      Array to be partitioned
   * @param low      Starting index
   * @param high     Ending index
   * @param strategy Pivot selection strategy
   * @return Final position of the pivot
   */
  private static int partition(int[] arr, int low, int high, PivotStrategy strategy) {
    // Select pivot based on strategy
    int pivotIndex = selectPivot(arr, low, high, strategy);

    // Move pivot to the end
    swap(arr, pivotIndex, high);
    int pivot = arr[high];

    // Index of smaller element
    int i = low - 1;

    for (int j = low; j < high; j++) {
      // If current element is smaller than or equal to pivot
      if (arr[j] <= pivot) {
        i++;
        swap(arr, i, j);
      }
    }

    // Place pivot in its correct position
    swap(arr, i + 1, high);
    return i + 1;
  }

  /**
   * Selects pivot index based on the chosen strategy.
   * 
   * @param arr      Array
   * @param low      Starting index
   * @param high     Ending index
   * @param strategy Pivot selection strategy
   * @return Index of selected pivot
   */
  private static int selectPivot(int[] arr, int low, int high, PivotStrategy strategy) {
    switch (strategy) {
      case FIRST:
        return low;

      case MIDDLE:
        return low + (high - low) / 2;

      case RANDOM:
        return low + (int) (Math.random() * (high - low + 1));

      case MEDIAN_OF_THREE:
        return medianOfThree(arr, low, high);

      case LAST:
      default:
        return high;
    }
  }

  /**
   * Finds the median of first, middle, and last elements.
   * 
   * @param arr  Array
   * @param low  Starting index
   * @param high Ending index
   * @return Index of median element
   */
  private static int medianOfThree(int[] arr, int low, int high) {
    int mid = low + (high - low) / 2;

    if (arr[low] > arr[mid]) {
      swap(arr, low, mid);
    }
    if (arr[low] > arr[high]) {
      swap(arr, low, high);
    }
    if (arr[mid] > arr[high]) {
      swap(arr, mid, high);
    }

    return mid;
  }

  /**
   * Sorts an array in descending order using quick sort.
   * 
   * @param arr Array of integers to be sorted
   * @return Sorted array in descending order
   */
  public static int[] quickSortDescending(int[] arr) {
    return quickSortDescending(arr, PivotStrategy.LAST);
  }

  /**
   * Sorts an array in descending order with specified pivot strategy.
   * 
   * @param arr      Array of integers to be sorted
   * @param strategy Pivot selection strategy
   * @return Sorted array in descending order
   */
  public static int[] quickSortDescending(int[] arr, PivotStrategy strategy) {
    if (arr.length <= 1) {
      return arr;
    }

    quickSortDescendingHelper(arr, 0, arr.length - 1, strategy);
    return arr;
  }

  /**
   * Helper method for descending quick sort.
   */
  private static void quickSortDescendingHelper(int[] arr, int low, int high, PivotStrategy strategy) {
    if (low < high) {
      int pivotIndex = partitionDescending(arr, low, high, strategy);
      quickSortDescendingHelper(arr, low, pivotIndex - 1, strategy);
      quickSortDescendingHelper(arr, pivotIndex + 1, high, strategy);
    }
  }

  /**
   * Partitions the array in descending order.
   */
  private static int partitionDescending(int[] arr, int low, int high, PivotStrategy strategy) {
    int pivotIndex = selectPivot(arr, low, high, strategy);
    swap(arr, pivotIndex, high);
    int pivot = arr[high];

    int i = low - 1;

    for (int j = low; j < high; j++) {
      if (arr[j] >= pivot) {
        i++;
        swap(arr, i, j);
      }
    }

    swap(arr, i + 1, high);
    return i + 1;
  }

  /**
   * Swaps two elements in an array.
   * 
   * @param arr Array
   * @param i   First index
   * @param j   Second index
   */
  private static void swap(int[] arr, int i, int j) {
    int temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
  }

  /**
   * Utility method to print an array
   */
  public static void printArray(int[] arr) {
    for (int i = 0; i < arr.length; i++) {
      System.out.print(arr[i] + " ");
    }
    System.out.println();
  }

  /**
   * Main method with example usage
   */
  public static void main(String[] args) {
    System.out.println("Quick Sort Examples:");

    // Test ascending sort with default pivot (LAST)
    int[] arr1 = { 64, 34, 25, 12, 22, 11, 90 };
    System.out.print("\nOriginal array: ");
    printArray(arr1);

    quickSort(arr1);
    System.out.print("Sorted array (ascending, LAST pivot): ");
    printArray(arr1);

    // Test with MEDIAN_OF_THREE pivot strategy
    int[] arr2 = { 64, 34, 25, 12, 22, 11, 90 };
    quickSort(arr2, PivotStrategy.MEDIAN_OF_THREE);
    System.out.print("Sorted array (MEDIAN_OF_THREE pivot): ");
    printArray(arr2);

    // Test with RANDOM pivot strategy
    int[] arr3 = { 64, 34, 25, 12, 22, 11, 90 };
    quickSort(arr3, PivotStrategy.RANDOM);
    System.out.print("Sorted array (RANDOM pivot): ");
    printArray(arr3);

    // Test descending sort
    int[] arr4 = { 64, 34, 25, 12, 22, 11, 90 };
    quickSortDescending(arr4);
    System.out.print("Sorted array (descending): ");
    printArray(arr4);

    // Test with already sorted array
    int[] arr5 = { 1, 2, 3, 4, 5 };
    System.out.print("\nAlready sorted array: ");
    printArray(arr5);
    quickSort(arr5, PivotStrategy.MEDIAN_OF_THREE);
    System.out.print("After quick sort: ");
    printArray(arr5);

    // Test with single element
    int[] arr6 = { 42 };
    System.out.print("\nSingle element array: ");
    printArray(arr6);
    quickSort(arr6);
    System.out.print("After quick sort: ");
    printArray(arr6);
  }
}

File: TypeScript/sorting/efficient/quickSort.ts
/**
 * Quick Sort Algorithm
 * 
 * Time Complexity: O(n log n) average, O(n²) worst case
 * Space Complexity: O(log n) due to recursion
 * 
 * Divide-and-conquer algorithm that selects a pivot element and partitions
 * the array around the pivot.
 */
class QuickSort {
  /**
   * Sorts an array using Quick Sort with last element as pivot (ascending order).
   * 
   * @param arr Array of numbers to be sorted
   * @return Sorted array
   */
  static sort(arr: number[]): number[] {
    if (arr.length <= 1) {
      return arr;
    }

    return this.quickSort(arr, 0, arr.length - 1);
  }

  /**
   * Quick Sort with first element as pivot.
   */
  static sortWithFirstPivot(arr: number[]): number[] {
    if (arr.length <= 1) {
      return arr;
    }

    return this.quickSortFirstPivot(arr, 0, arr.length - 1);
  }

  /**
   * Quick Sort with middle element as pivot.
   */
  static sortWithMiddlePivot(arr: number[]): number[] {
    if (arr.length <= 1) {
      return arr;
    }

    return this.quickSortMiddlePivot(arr, 0, arr.length - 1);
  }

  /**
   * Quick Sort with random pivot.
   */
  static sortWithRandomPivot(arr: number[]): number[] {
    if (arr.length <= 1) {
      return arr;
    }

    return this.quickSortRandomPivot(arr, 0, arr.length - 1);
  }

  /**
   * Recursive quick sort (last element as pivot).
   */
  private static quickSort(arr: number[], low: number, high: number): number[] {
    if (low < high) {
      const pi = this.partitionLast(arr, low, high);

      this.quickSort(arr, low, pi - 1);
      this.quickSort(arr, pi + 1, high);
    }

    return arr;
  }

  /**
   * Recursive quick sort (first element as pivot).
   */
  private static quickSortFirstPivot(arr: number[], low: number, high: number): number[] {
    if (low < high) {
      const pi = this.partitionFirst(arr, low, high);

      this.quickSortFirstPivot(arr, low, pi - 1);
      this.quickSortFirstPivot(arr, pi + 1, high);
    }

    return arr;
  }

  /**
   * Recursive quick sort (middle element as pivot).
   */
  private static quickSortMiddlePivot(arr: number[], low: number, high: number): number[] {
    if (low < high) {
      const pi = this.partitionMiddle(arr, low, high);

      this.quickSortMiddlePivot(arr, low, pi - 1);
      this.quickSortMiddlePivot(arr, pi + 1, high);
    }

    return arr;
  }

  /**
   * Recursive quick sort (random pivot).
   */
  private static quickSortRandomPivot(arr: number[], low: number, high: number): number[] {
    if (low < high) {
      const pi = this.partitionRandom(arr, low, high);

      this.quickSortRandomPivot(arr, low, pi - 1);
      this.quickSortRandomPivot(arr, pi + 1, high);
    }

    return arr;
  }

  /**
   * Partition using last element as pivot.
   */
  private static partitionLast(arr: number[], low: number, high: number): number {
    const pivot = arr[high];
    let i = low - 1;

    for (let j = low; j < high; j++) {
      if (arr[j] < pivot) {
        i++;
        [arr[i], arr[j]] = [arr[j], arr[i]];
      }
    }

    [arr[i + 1], arr[high]] = [arr[high], arr[i + 1]];
    return i + 1;
  }

  /**
   * Partition using first element as pivot.
   */
  private static partitionFirst(arr: number[], low: number, high: number): number {
    // Move first element to end and use regular partition
    [arr[low], arr[high]] = [arr[high], arr[low]];
    return this.partitionLast(arr, low, high);
  }

  /**
   * Partition using middle element as pivot.
   */
  private static partitionMiddle(arr: number[], low: number, high: number): number {
    const mid = Math.floor((low + high) / 2);
    [arr[mid], arr[high]] = [arr[high], arr[mid]];
    return this.partitionLast(arr, low, high);
  }

  /**
   * Partition using random pivot.
   */
  private static partitionRandom(arr: number[], low: number, high: number): number {
    const randomIdx = Math.floor(Math.random() * (high - low + 1)) + low;
    [arr[randomIdx], arr[high]] = [arr[high], arr[randomIdx]];
    return this.partitionLast(arr, low, high);
  }

  /**
   * Sorts an array in descending order.
   */
  static sortDescending(arr: number[]): number[] {
    if (arr.length <= 1) {
      return arr;
    }

    this.sort(arr);
    return arr.reverse();
  }
}
